lib/bap_disasm/bap_disasm_symtab.ml

```diff
@@ -52,19 +52,27 @@ let merge m1 m2 =
   Memmap.to_sequence m2 |> Seq.fold ~init:m1 ~f:(fun m1 (mem,x) ->
       Memmap.add m1 mem x)
 
+let filter_mem mem name entry =
+  Memmap.filter mem ~f:(fun (n,e,_) ->
+      not(String.(name = n) || Block.(entry = e)))
+
 let remove t (name,entry,_) : t = {
   names = Map.remove t.names name;
   addrs = Map.remove t.addrs (Block.addr entry);
-  memory = Memmap.filter t.memory ~f:(fun (n,e,_) ->
-      not(String.(name = n) || Block.(entry = e)))
+  memory = filter_mem t.memory name entry;
 }
 
+let filter t ((name,entry,_ ) as fn) =
+  if Map.mem t.names name || Map.mem t.addrs (Block.addr entry) then
+    remove t fn
+  else t
+
 let add_symbol t (name,entry,cfg) : t =
   let data = name,entry,cfg in
-  let t = remove t data in
+  let t = filter t data in
```

Comment:
can you please justify why we can use `filter` instead of `remove` here? It's not quite obvious to me.